File: java/annalyns-infiltration/src/main/java/AnnalynsInfiltration.java
class AnnalynsInfiltration {
    public static boolean canFastAttack(boolean knightIsAwake) {
      return !knightIsAwake;

    }

    public static boolean canSpy(boolean knightIsAwake, boolean archerIsAwake, boolean prisonerIsAwake) {
        return (knightIsAwake || archerIsAwake || prisonerIsAwake);


    }

    public static boolean canSignalPrisoner(boolean archerIsAwake, boolean prisonerIsAwake)  {
      return (!archerIsAwake && prisonerIsAwake);

    }

    public static boolean canFreePrisoner(boolean knightIsAwake, boolean archerIsAwake, boolean prisonerIsAwake, boolean petDogIsPresent) {
        return (petDogIsPresent && !archerIsAwake) || (prisonerIsAwake && !knightIsAwake && !archerIsAwake);

        /*
        if(archerIsAwake)
        {
            if(!prisonerIsAwake && !knightIsAwake &&!petDogIsPresent)
            {
                return false;
            }
        }
        else
            return true;
        throw new UnsupportedOperationException("Please implement the (static) AnnalynsInfiltration.canFreePrisoner() method");
    }*/
    }
}

File: csharp/annalyns-infiltration/AnnalynsInfiltration.cs
using System;

static class QuestLogic
{
      public static bool CanFastAttack(bool knightIsAwake)
    {
        return !knightIsAwake;
    }

    public static bool CanSpy(bool knightIsAwake, bool archerIsAwake, bool prisonerIsAwake)
    {
       if(knightIsAwake || archerIsAwake || prisonerIsAwake)
           return true;
        return false;
    }

    public static bool CanSignalPrisoner(bool archerIsAwake, bool prisonerIsAwake)
    {
        return (!archerIsAwake && prisonerIsAwake);
    }

    public static bool CanFreePrisoner(bool knightIsAwake, bool archerIsAwake, bool prisonerIsAwake, bool petDogIsPresent)
    {
       if(petDogIsPresent && !archerIsAwake)
           return true;
        if(!petDogIsPresent && prisonerIsAwake)
            if(!knightIsAwake && !archerIsAwake)
                return true;
        return false;
    }
}
